### PR TITLE
Add explicit header mutation validation ext_authz

### PR DESCRIPTION
**Pull request by antoniovleonti**
Commit Message: Add explicit header mutation validation ext_authz
Additional Description: Currently, neither headers nor query parameters from the sidestream are validated within ext_authz. This causes two issues:

1. ext_authz's fuzzers cannot fuzz responses easily since the fuzzer likes to add illegal characters to string fields. These will fail because header keys get converted to a LowerCaseString, which will cause an ASSERT to fail within LowerCaseString.
2. ASSERTs do nothing in production, so if the external auth service makes invalid header mutations (e.g. adds a header with an invalid key), these mutations will be applied and will be visible to other filters. (Headers are validated before they get encoded, so they will not be sent upstream or downstream.) Likewise, if the sidestream adds invalid query parameters, the path for the request will become invalid as well.

This PR adds explicit validation to ext_authz. It also adds percent-encoding for query parameters to ensure that the resulting path is always valid. Note that headers are checked for validity and either added as they came or ignored, while query parameters are always encoded & added.

Also, while updating ext_authz_test.cc, I also noticed that `HttpFilterTestParam`'s parameters were not used anywhere, so I fixed that.

Risk Level: Low
Testing: unit tested, not easy to write an integration test for since it's hard to check a HeaderMap for a header whose key is invalid (since it fails LowerCaseString validation)
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by antoniovleonti** [comment]
I should validate query parameters as well since those are used to update the path pseudoheader and thus can cause another ASSERT to fail.

**Comment by soulxu** [comment]
cc @esmet @tyxia @ggreenway

**Comment by tyxia** [comment]
/assign @yanjunxiang-google 

Could you also take a look when you have time? Thanks!

File: source/extensions/filters/common/ext_authz/ext_authz.h
Comment (yanjunxiang-google) [inline source/extensions/filters/common/ext_authz/ext_authz.h:76]: Thanks for working on this!

Just a high level comment first:
It looks to me you are copying the response header as it is into ext_authz data structure, then trying to validate it when ApplyHeaderMutation().  I am wondering can you do the validation before copying the response header into ext_authz data structure. That way, you don't need to modify these data types in ext_authz.h.

Comment (antoniovleonti) [reply source/extensions/filters/common/ext_authz/ext_authz.h:76]: You are correct, that is what I'm doing. I considered doing it in a different way that didn't require changing this header, but I don't like the idea of adding the same validation code to both clients.

Comment (yanjunxiang-google) [reply source/extensions/filters/common/ext_authz/ext_authz.h:76]: @ggreenway @tyxia do you have some comments on this change?

File: source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc
Comment (yanjunxiang-google) [inline source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc:55]: Is it simpler to pass a reference of ok_response as multiple parameters are within it?

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (yanjunxiang-google) [inline source/extensions/filters/http/ext_authz/ext_authz.cc:320]: Just curious why we are doing this. Same as below. places.

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (yanjunxiang-google) [inline source/extensions/filters/http/ext_authz/ext_authz.cc:449]: Is adding headerPairIsValid() to different type of the headers, like headers_to_set, headers_to_add, headers_to_append, headers_to_remove the main change of the PR? And everything else is just refactoring to make it a center place to do this?

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (antoniovleonti) [reply source/extensions/filters/http/ext_authz/ext_authz.cc:320]: Because query parameters get added to the path header, so if the query parameters are incorrectly encoded, the resulting path will also be incorrectly encoded.

Some kind of encoding / validation here is necessary, if for no other reason than adding CheckResponse to the ext_authz fuzzer input. The fuzzer will add invalid characters to the query parameters, which will then get added to the path, which will then cause an assert failure.

File: source/extensions/filters/common/ext_authz/ext_authz.h
Comment (yanjunxiang-google) [reply source/extensions/filters/common/ext_authz/ext_authz.h:76]: Understood there might be a little bit code duplication.  However, if you do it early on, i.e, in the client, you are able to avoid unnecessary header copy and storage, right?

File: source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc
Comment (antoniovleonti) [reply source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc:55]: I've written it so that it uses the same function for copying denied_response and ok_response header mutations. denied_response does not use all the same fields.

File: source/extensions/filters/common/ext_authz/ext_authz.h
Comment (antoniovleonti) [reply source/extensions/filters/common/ext_authz/ext_authz.h:76]: I think this only causes more copies for the `response_headers_to_...` fields, since instead of moving the entire vector, it only copies the valid headers. For all other fields, there are the same amount of copies: one copy into the "internal" Response struct, then again into the HeaderMap. 

I think there may be room to optimize the copies, for example by cannibalizing the response proto in the grpc client's onSuccess.

@tyxia @ggreenway @esmet Thoughts?

Comment (antoniovleonti) [reply source/extensions/filters/common/ext_authz/ext_authz.h:76]: Hm. On second thought, the validation might only be necessary in the grpc client anyway. In the raw http client, the headers are being sent over http, so added headers should always be valid.

One exception is that the side stream could add an invalid header to remove (since it's sent as a header VALUE), but that is easy enough to validate.

Comment (yanjunxiang-google) [reply source/extensions/filters/common/ext_authz/ext_authz.h:76]: That's a good point.  For HTTP response received by ext_authz HTTP client from the side stream, whether we should validate is a good question.  I assume Envoy HTTP codec/parser already validated them.  Please confirm this.

File: test/extensions/filters/common/ext_authz/ext_authz_grpc_impl_test.cc
Comment (yanjunxiang-google) [inline test/extensions/filters/common/ext_authz/ext_authz_grpc_impl_test.cc:132]: Do we have some tests for HTTP client?

File: test/extensions/filters/http/ext_authz/ext_authz_test.cc
Comment (yanjunxiang-google) [inline test/extensions/filters/http/ext_authz/ext_authz_test.cc:480]: You mean "Add a invalid header" in the comment?
Same as below

File: test/extensions/filters/http/ext_authz/ext_authz_test.cc
Comment (yanjunxiang-google) [inline test/extensions/filters/http/ext_authz/ext_authz_test.cc:265]: Do you need to add a test to verify the invalid headers {"invalid-\nkey", "bar"} is not added? Or you already have such test somewhere?

**Comment by yanjunxiang-google** [comment]
I finished my first round pass on this PR. @antoniovleonti when you are ready, please engage the ext_authz code owners for reviewing:  https://github.com/envoyproxy/envoy/blob/313b6fb7cf0f806e74a2d42c93e7c1fcccce2391/CODEOWNERS#L279

File: test/extensions/filters/common/ext_authz/ext_authz_grpc_impl_test.cc
Comment (antoniovleonti) [reply test/extensions/filters/common/ext_authz/ext_authz_grpc_impl_test.cc]: Done

File: source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc
Comment (tyxia) [inline source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc:61]: Should we use runtime guard for the change of using URL-encoding? This will introduces the behavior change.

Reference in the past: `envoy_reloadable_features_oauth_use_url_encoding`

**Review by tyxia — COMMENTED** [review]
LGTM modulo runtime gard.

Thanks!

File: source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.h
Comment (tyxia) [inline source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.h:65]: One thing to confirm: The validation (which is introduced by this function) is only applied to gRPC client, right?

File: source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc
Comment (ggreenway) [inline source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc:60]: The auth service is sending back an invalid url parameter. Would it be better to drop it, and increment a metric, instead of url-encoding? I'd argue that the service should've already url-encoded it if that was the intention.

What is the current end-to-end behavior if the service sends back a query param that should have been url-encoded? Where do things fail?

Either way, this will need either a runtime guard or config for the behavior change.

File: source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc
Comment (ggreenway) [inline source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc:116]: I think silently dropping requested mutations is a mistake. I'd rather fail the entire client request in this case. 

A runtime guard on the behavior change (or config) is needed.

**Review by ggreenway — CHANGES_REQUESTED** [review]
/wait

File: source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc
Comment (antoniovleonti) [reply source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc:116]: Failing on invalid mutations is also possible, but the side stream has already indicated that the client request should be accepted. Perhaps it should be configurable?

File: source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.h
Comment (antoniovleonti) [reply source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.h:65]: No, the http client validates headers to remove.

File: source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc
Comment (antoniovleonti) [reply source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc:116]: I agree with your reasoning!

File: source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc
Comment (antoniovleonti) [reply source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc:60]: For clarity: when you say "drop it," you mean drop the client request, yes?

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_authz/ext_authz.cc]: Is there a bug in the original code which can justify the URL encoding change here?
If not, can we just do normal string checking here?

File: source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc
Comment (antoniovleonti) [reply source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc:60]: I have not tested invalid query parameter behavior, but I know invalid headers are rejected at the codec.

Invalid query parameters would make the path pseudoheader invalid, so I think it would be similar.

I'll test this tomorrow.

File: source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc
Comment (tyxia) [reply source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc:116]: Reject the whole request is good, but I think we should make sure such rejection caused by invalid mutation is different from existing rejection caused by authorization failure. (e.g., the response code and responses body in localReply should be different)

Comment (ggreenway) [reply source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc:116]: Agreed; the response code should be in the 5xx range, because this is a server-side bug in the auth service.

File: source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.h
Comment (antoniovleonti) [reply source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.h:65]: 1. In the http client, headers to add / append are expressed via header names over raw http. If the side stream were to express an invalid header in this way, they would just create an invalid http response.
2. I added validation for headers_to_remove because it gets turned into a LowerCaseString. If LowerCaseString::valid fails, it causes an assertion failure. Thus we cannot fuzz the side stream response.

I'm fine with not adding validation to the http client in this PR. The main thing I want to get out of this is fuzzing the grpc response.

Comment (antoniovleonti) [reply source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.h:65]: http client changes have been removed from this PR.

File: source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc
Comment (antoniovleonti) [reply source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc]: Done.

File: source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc
Comment (antoniovleonti) [reply source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc]: Done.

File: api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto
Comment (wbpcode) [inline api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto:98]: nit:  change name to `status_on_validation_failure`

---

Sorry, seems I am involved because this API change. But rather than a runtime guard that will finally be deprecated and removed, I prefer an explicitly API to enable the headers validation would better.
I believe most users will use a trusted auth server and even some users may doesn't want to reject invalid headers (even HCM could accept invalid header name with '_' by setting).

/wait-any

Comment (antoniovleonti) [reply api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto:98]: Thanks for that insight, that makes sense.

File: source/extensions/filters/common/ext_authz/ext_authz.h
Comment (antoniovleonti) [reply source/extensions/filters/common/ext_authz/ext_authz.h]: This had the unintended side effect of adding validation to the network/ext_authz filter so I'm going to move the validation code back to the filter.

The network/ext_authz filter doesn't read or care about header / query parameter mutations, so it makes no sense for the client to ever reject the response because them. The mutation validation is only needed in the http filter, so the code should go in the http filter.

File: source/extensions/filters/common/ext_authz/ext_authz.h
Comment (ggreenway) [inline source/extensions/filters/common/ext_authz/ext_authz.h:77]: I think I'd prefer to name this `Invalid`.

File: api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto
Comment (ggreenway) [inline api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto:98]: Does this need it's own config? Can't this just use `status_on_error`? I don't really see any difference between getting an error from the auth server and getting an invalid response.

Comment (ggreenway) [reply api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto:98]: Agreed; this makes more sense as config than a runtime guard.

**Review by ggreenway — CHANGES_REQUESTED** [review]
/wait

File: api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto
Comment (antoniovleonti) [reply api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto:98]: > @[tyxia](https://github.com/tyxia) tyxia [last week](https://github.com/envoyproxy/envoy/pull/33244#discussion_r1577055029) • 
> Rejecting the whole request is good, but I think we should make sure such rejection caused by invalid mutation is different from existing rejection caused by authorization failure. (e.g., the response code and responses body in localReply should be different)

> @[ggreenway](https://github.com/ggreenway) ggreenway [last week](https://github.com/envoyproxy/envoy/pull/33244#discussion_r1577072899)
> Agreed; the response code should be in the 5xx range, because this is a server-side bug in the auth service.

By default `status_on_error` is 403 (forbidden). I do think internal server error is a more appropriate status.

Comment (ggreenway) [reply api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto:98]: Oh right, I remember that now. Yes, this makes sense. Although I'm still not sure it needs config; it could always be 500 (internal server error).

Comment (antoniovleonti) [reply api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto:98]: Fine by me, I was just copying the pattern I saw for the error code

File: test/extensions/filters/http/ext_authz/ext_authz_integration_test.cc
Comment (ggreenway) [inline test/extensions/filters/http/ext_authz/ext_authz_integration_test.cc:1103]: Maybe add `test_server_->waitForCounterGe("some.path.ext_authz.invalid", 1);`

File: test/extensions/filters/http/ext_authz/ext_authz_test.cc
Comment (ggreenway) [inline test/extensions/filters/http/ext_authz/ext_authz_test.cc:59]: Please revert. It adds a lot of noise to the diff.

**Review by ggreenway — CHANGES_REQUESTED** [review]
This looks great overall!

/wait

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (antoniovleonti) [reply source/extensions/filters/http/ext_authz/ext_authz.cc:384]: Sorry, I meant to make this config guarded like the other behavior changes. Nice catch.

File: source/extensions/filters/http/ext_authz/ext_authz.cc
Comment (antoniovleonti) [reply source/extensions/filters/http/ext_authz/ext_authz.cc:468]: I was inconsistent with how I formatted these kinds of conditions; they all look like how you've suggested now.

**Comment by ggreenway** [comment]
@wbpcode please review on behalf of api-shephards

File: source/extensions/filters/http/ext_authz/ext_authz.h
Comment (nezdolik) [reply source/extensions/filters/http/ext_authz/ext_authz.h:109]: don't think this ancient TODO has ever been addressed, but the idea was to move those stats from cluster scope to filter scope. Some more context here: https://github.com/envoyproxy/envoy/pull/8436#discussion_r335592361. Looks like we still populate cluster level stats in the filter: https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/http/ext_authz/ext_authz.cc#L429

**Pull request merged**
